nemo/core/neural_modules.py

```diff
@@ -438,89 +561,113 @@ def __call__(self, **kwargs):
         Returns:
           NmTensor object or tuple of NmTensor objects
         """
-        # Get input and output ports definitions.
-        input_port_defs = self.input_ports
-        output_port_defs = self.output_ports
-
-        first_input_nmtensor_type = None
-        input_nmtensors_are_of_same_type = True
-        for port_name, tgv in kwargs.items():
-            # make sure that passed arguments correspond to input port names
-            if port_name not in input_port_defs.keys():
-                raise NeuralPortNameMismatchError("Wrong input port name: {0}".format(port_name))
-
-            input_port = input_port_defs[port_name]
-            type_comatibility = input_port.compare(tgv)
-            if (
-                type_comatibility != NeuralTypeComparisonResult.SAME
-                and type_comatibility != NeuralTypeComparisonResult.GREATER
-            ):
-                raise NeuralPortNmTensorMismatchError(
-                    "\n\nIn {0}. \n"
-                    "Port: {1} and a NmTensor it was fed are \n"
-                    "of incompatible neural types:\n\n{2} \n\n and \n\n{3}"
-                    "\n\nType comparison result: {4}".format(
-                        self.__class__.__name__, port_name, input_port_defs[port_name], tgv, type_comatibility,
-                    )
+        # print(" Neural Module:__call__")
+
+        # Set the operation mode of the outer graph.
+        self.operation_mode = self._app_state.active_graph.operation_mode
+        # The input and output ports definitions can potentially depend on the operation mode!
+
+        # Record the operation (i.e. add a single module).
+        step_number = self._app_state.active_graph.record_step(self)
+
+        ###### PROCESS INPUTS. ######
+        # Iterate through all passed parameters.
+        for port_name, port_content in kwargs.items():
+            # Make sure that passed arguments corresponds to one of the input port names.
+            if port_name not in self.input_ports.keys():
+                raise NeuralPortNameMismatchError(port_name)
+
+            # At that point the input can be one of three types:
+            # * NeuralGraph -> bind port using the default name and type.
+            # * GraphInput -> check definition, if ok bind port.
+            # * NmTensor -> check definition, add self as a "consumer" of a tensor (produced by other module).
+
+            # Check what was actually passed.
+            if type(port_content).__name__ == "NeuralGraph":
```

Comment:
Can you explain when you would pass an NeuralGraph to a NeuralModule?

Reply:
For the "binding of input port without renaming" purpose.
Please take a look at the table in "Neural Graph Summary", 5. Bounding of Graph inputs and outputs:
https://docs.google.com/document/d/1t1H-PYNjEFgc_BmSN-R8yXq8u_qykHEYPVK5cH7aaPM/edit#

Reply:
We can consider this resolved.
However, I do want to note that I consider this option to be unintuitive.

Reply:
Ok, like I said - it it is unintuitive then the users won't use it and it will die with "natural death" ;)

And I like the idea of having this:
with g1:
  # Bind the input port using the same name of port and type.
  y_pred = tn(x=g1.**inputs**)
  # Result: Input g1[“x”] is “connected” to tn.input_ports[“x”]

Actually I am working on it, we'll see how much time it will take, in theory all bricks are down there already...